Validator list

kylegranger opened the pull request:
This code (my very first rust code ever) basically used this rippled function  -- `buildValidatorList()` -- as a model, to see what steps are used to create and sign a manifest, as well as a validator blob.
https://github.com/XRPLF/rippled/blob/master/src/test/overlay/compression_test.cpp#L316

The verification side does the following:
- receives a manifest
- extracts the signatures
- recreates the manifest without the signatures
- appends a 4-char prefix to it, 
- gets a 32-byte digest from SHA512 (throwing away half of those 64 bytes),
- then does the signature verification, once for each of three different signature (matching each sign() call in the C++ code), also with the validator blob data.

Our keys are generated before hand, one key pair for the master key, and one for the signing key.  The master public key has been added in `validators.txt`.  The readme has been update to reflect this new content.

Rqnsom commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +71,259 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+
+    // 3. append manifest prefix
+    let mut prefixed_signable: Vec<u8> = vec![0; signable_manifest.len() + 4];
+    prefixed_signable[0..4].clone_from_slice(man_prefix.as_slice());
```
Would maybe this do the trick? Then the `man_prefix` variable is not needed at all.
```suggestion
    prefixed_signable[0..4].clone_from_slice(b"MAN\x00");
```

Rqnsom reviewed (commented):
I have to go afk, but I will check more later, but here are some comments I've written so far

asmie commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +71,259 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+
+    // we return 32 bytes of 64-byte result
+    let mut signature = [0u8; 32];
+    for i in 0..32 {
```
Instead of the loop it could be possible to do this:
`signature.copy_from_slice(&result[..32]);`

kylegranger commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +71,259 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+        .unicast(node.addr(), payload)
+        .expect("unable to send message");
+
+    sleep(Duration::from_secs(300)).await;
```
The expectation is to (eventually) check the response that the message was accepted by rippled, and not invalid.  Lukasz and I decided to do this in a second follow up task/PR.
I changed the sleep to 5 seconds, enough time for me to confirm that the message is accepted by rippled (manually checking the console output).

Rqnsom replied:
> The expectation is to (eventually) check the response that the message was accepted by rippled, and not invalid. Lukasz and I decided to do this in a second follow up task/PR.

aha, that makes sense. Could we leave a TODO comment then?

lwojtow commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +71,255 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+
+fn create_final_manifest(
+    sequence: u32,
+    public_key: &Vec<u8>,
```
Whenever you pass &Vec<T>, you could pass &[T]. This will save one pointer derefence (and I think clippy should complain about it).

lwojtow commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -1,27 +1,43 @@
 #[allow(non_snake_case)]
 async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
     // ZG-CONFORMANCE-015
-
```
Please add this line back.

lwojtow reviewed (commented):
I realize this is your first rust code, so forgive me being picky about style. Nevertheless, this is fantastic piece of work :)

Rqnsom commented on README.md:
````diff
@@ -6,10 +6,86 @@ The Ziggurat implementation for XRPLF's `rippled` nodes.
-    ```
-    curl --create-dirs --output ~/.ziggurat/ripple/setup/validators.txt https://raw.githubusercontent.com/XRPLF/rippled/develop/cfg/validators-example.txt
-    ```
+3. Create the `~/.ziggurat/ripple/setup` directories. Next, create a file there named `validators.txt` with the following content:
+
+```
+#
+# Default validators.txt
+#
+# This file is located in the same folder as your rippled.cfg file
+# and defines which validators your server trusts not to collude.
+#
+# This file is UTF-8 with DOS, UNIX, or Mac style line endings.
+# Blank lines and lines starting with a '#' are ignored.
+#
+#
+#
+# [validators]
+#
+#   List of the validation public keys of nodes to always accept as validators.
+#
````
General comment, I think this can be a separate PR:

These instructions are getting really huge. I would move lots of stuff here into a script, like we did in Algorand project. Check how instructions are simple here: https://github.com/runziggurat/algorand#getting-started

Alternative is:
This file could be put into [xrpl](https://github.com/runziggurat/xrpl)/setup/ directory,
and then we can have a `cp setup/validators.txt ~/.ziggurat......` command here

kylegranger replied:
I liked the idea of file in `xrpl/setup`.  Did that, and updated readme accordingly.

Rqnsom commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +71,255 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+
+#[tokio::test]
+#[allow(non_snake_case)]
+async fn c026_TM_VALIDATOR_LIST_send_validator_list() {
```
I think you may have missed updating the SPEC.md document.

Usually every test we have in `.rs` files has a description in the [SPEC.md](https://github.com/runziggurat/xrpl/blob/main/SPEC.md) document

lwojtow replied:
As much as I agree with keeping SPEC.md up to date, I think we can leave it until the test is finished. (Or maybe just put some 'placeholder' kind of comment there for now).

kylegranger commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +71,255 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+    signing_pub_key: &Vec<u8>,
+) -> Vec<u8> {
+    let size = 5 + 2 + public_key.len() + 2 + signing_pub_key.len();
+    let mut manifest: Vec<u8> = vec![0; size];
```
I've cleaned up the manifest serialization quite a bit already, but yeah, I'll go ahead and jump with `bytes`.  I think it's the right way to do this, and best to bite the bullet sooner rather than later.

kylegranger commented:
last of changes have been committed, including:
- serialization uses bytes::BytesMut
- all magic numbers moved to consts
- validators.txt file added to project, in setup folder
- readme updated, for copying this file to ~./ziggurat/...
- use &str over String, pass &[T] over &Vec<T>
- shorten sleep from 5 min to 5 secs
- test renamed, updated SPEC.md, added TODO for real test verification.
- use string literal for 4-char u32
- removed unused code, moved expiration creation to own function

niklaslong commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +85,212 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+    let mstr = base64::encode(manifest);
+    let mb = mstr.as_bytes().to_vec();
+    let sstr = hex::encode_upper(blob_signature_bytes);
+    let sb = sstr.as_bytes().to_vec();
```
Small nit/suggestion: perhaps we could name these variables more explicitly, a few chars less isn't necessarily worth the loss in "understandable at a glance" quality?

Rqnsom commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +85,212 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+    // 5. Sign it with signing private key, get signature
+    let signature_bytes = sign_buffer(&signing_secret_key, &prefixed_signable);
+
+    // 6. Create final manifest with sequence, public key, signing public key, master signature, signature
+    let manifest = create_final_manifest(
+        1,
```
General improvement 2:

Just like they did in [their code](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/impl/Sign.cpp#L25), we could handle the prefix from the `sign_buffer` function, e.g.:
```
let master_signature_bytes = sign_buffer(&hash_prefix, &master_secret_key, &manifest);
// or just call it `sign` instead of `sign_buffer` to be aligned with their code base
```


------
Then the current `create_final_manifest` function could be renamed to `sign_manifest`, and we could reuse the old manifest here (also make the `create_final_manifest` function smaller since we would only need to append `master_signature_bytes` and `signature_bytes` from that function to our already existing `manifest`):
```
sign_manifest(&mut manifest, &master_signature_bytes, &signature_bytes);
```

Rqnsom reviewed (commented):
I finally managed to get through this PR, it wasn't a straightforward one to understand,
but good work! Looks quite good, I left a few extra comments

kylegranger commented:
This should be pretty solid now.  Last round of changes included:
- refactored creating manifest, signing, appending prefix, signing buffers
- renamed many, many variables.
- restored text in `SPEC.md` mistakenly deleted (merge error)

niklaslong commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +86,196 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+    let manifest_b64 = base64::encode(signed_manifest);
+    let manifest_b64_bytes = manifest_b64.as_bytes().to_vec();
+    let siganture_hex = hex::encode_upper(signature);
+    let siganture_hex_bytes = siganture_hex.as_bytes().to_vec();
```
```suggestion
    let signature_hex_bytes = signature_hex.as_bytes().to_vec();
```

Rqnsom commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -239,48 +230,40 @@ async fn c026_TM_VALIDATOR_LIST_send_validator_list() {
+    //let mut prefixed_signable = BytesMut::with_capacity(1024);
+
+    // 3. Sign the signable manifest with master secret key, get master signature
+    let master_signature_bytes =
+        sign_buffer_with_prefix(MANIFEST_PREFIX, &master_secret_key, &manifest);
```
I still think we can avoid this extra function which handles the `prefix` case

```suggestion
    let master_signature_bytes =
        sign_buffer(Some(MANIFEST_PREFIX), &master_secret_key, &manifest);
```

Then on the line below, we can reuse the function like this:
```rust
    // 8.  Get signature for blob using master private key
    let signature = sign_buffer(None, &signing_secret_key, &blob_bytes);
```

kylegranger replied:
in my opinion.
- I think it's good to have a function that just does the signing -- I don't consider it 'extra'
- appending of a prefix is somewhat orthogonal to the signing.
- I don't like passing the idea of passing in 'None'.

Rqnsom commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +86,193 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+    // 5. Create signed manifest with sequence, public key, signing public key, master signature, signature
+    let signed_manifest = sign_manifest(&manifest, &master_signature_bytes, &signature_bytes);
+
+    // 6. Create Validator blob.
+    let blob = create_validator_blob_json(&signed_manifest, master_public_hex);
+    let blob_b64 = base64::encode(&blob);
+    let blob_bytes = base64::decode(&blob_b64).expect("unable to decode a blob");
```
Let's try to avoid doing a decode operation just after we encode something. We could instead copy the pre-encoded data into a separate variable before we do the encoding.
Or see my other suggestion below.

kylegranger replied:
done

Rqnsom reviewed (commented):
Looks pretty good

kylegranger commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +86,193 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+    let signing_public_hex = "03859B76317C8AA64F2D253D3547831E413F2663AE2568F7A17E85B283CC8861E4";
+    let signing_secret_bytes = hex::decode(signing_secret_hex).expect("unable to decode hex");
+    let signing_public_bytes = hex::decode(signing_public_hex).expect("unable to decode hex");
+    let signing_secret_key = SecretKey::from_slice(signing_secret_bytes.as_slice())
```
done

kylegranger commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +86,193 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+        blob: blob_b64_bytes,
+        signature: signature_hex_bytes,
+        version: 1,
+    });
```
done

kylegranger commented on src/tests/conformance/post_handshake/validators.rs:
```diff
@@ -55,3 +86,193 @@ async fn c015_TM_VALIDATOR_LIST_COLLECTION_node_should_send_validator_list() {
+
+    // copy manifest into buffer
+    buf.extend_from_slice(manifest);
+
```
done

Rqnsom reviewed (approved):
LGTM 🚀
